File: IntersectionOf2LinkedList.cpp
//Time Complexity - O(n)
//Space Complexity - O(1)

//Problem successfully compiled on leet code
//No problems faced while coding this.


#include<iostream>
using namespace std;


  struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };
 
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int lenA=0;
        int lenB=0;
        
        ListNode* A=headA;
        ListNode* B=headB;
        while(A!=NULL)
        {
            A=A->next;
            lenA++;
        }
        while(B!=NULL)
        {
            B=B->next;
            lenB++;
        }
        A=headA;
        B=headB;
        if(lenA>lenB)
        {
            int timesToGoFront=lenA-lenB;
            
            while(timesToGoFront>0)
            {
                A=A->next;
                timesToGoFront--;
            }
        }
        else if(lenB>lenA)
        {
            int timesToGoFront=lenB-lenA;
            while(timesToGoFront>0)
            {
                B=B->next;
                timesToGoFront--;
            }
        }
        
        
        while(A!=NULL)
        {
            if(A==B)
            {
                return A;
            }
            A=A->next;
            B=B->next;
        }
        
        return NULL;
        
    }
};

File: intersection of 2 linked list.py
# // Time Complexity : O(n+m) where n and m are the length of linked lists
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

class Solution:
    def getIntersectionNode(self, headA, headB):
        
        # set both pointers at the heads of both linked lists 
        a = headA
        b = headB
        
        # loop ends at the intersection
        while a != b:
            if not a:
                a = headB
            else:
                a = a.next
                
            if not b:
                b = headA
            else:
                b = b.next
                
        return a